compiler/hash-pipeline/src/lib.rs

```diff
@@ -15,127 +13,61 @@ pub mod traits;
 use std::{collections::HashMap, env, time::Duration};
 
 use fs::{read_in_path, resolve_path, PRELUDE};
-use hash_ast::{ast::OwnsAstNode, tree::AstTreeGenerator, visitor::AstVisitor};
 use hash_reporting::{report::Report, writer::ReportWriter};
 use hash_source::{constant::CONSTANT_MAP, ModuleKind, SourceId};
-use hash_utils::{path::adjust_canonicalisation, timing::timed, tree_writing::TreeWriter};
-use settings::{CompilerMode, CompilerSettings};
+use hash_utils::timing::timed;
+use settings::{CompilerSettings, CompilerStageKind};
 use sources::{Module, Workspace};
-use traits::{Desugar, Lowering, Parser, SemanticPass, Tc, VirtualMachine};
+use traits::CompilerStage;
 
 pub type CompilerResult<T> = Result<T, Vec<Report>>;
 
 /// The Hash Compiler interface. This interface allows a caller to create a
 /// [Compiler] with the specified components. This allows external tinkerers
 /// to add their own implementations of each compiler stage with relative ease
 /// instead of having to scratch their heads.
-pub struct Compiler<'pool, P, D, S, C, L, V> {
-    /// The parser stage of the compiler.
-    parser: P,
-    /// De-sugar the AST
-    desugarer: D,
-    /// Perform semantic analysis on the AST.
-    semantic_analyser: S,
-    /// The typechecking stage of the compiler.
-    checker: C,
-    /// The IR Lowerer.
-    lowerer: L,
-    /// The current VM.
-    vm: V,
+pub struct Compiler<'pool> {
+    /// The attached stages of the compiler pipeline.
+    stages: Vec<Box<dyn CompilerStage<'pool>>>,
+
     /// Various settings for the compiler.
     pub settings: CompilerSettings,
     /// The pipeline shared thread pool.
     pool: &'pool rayon::ThreadPool,
 
     /// A record of all of the stage metrics
-    metrics: HashMap<CompilerMode, Duration>,
+    metrics: HashMap<CompilerStageKind, Duration>,
```

Comment:
What if two different stages have the same kind?

Reply:
Multiple stages are allowed to have the same `CompilerStageKind`, and for the metrics, it will essentially be summed.